File: java/datastructures/lib/src/main/java/datastructures/trees/BinaryTree.java
package datastructures.trees;

import datastructures.queue.Queue;
import datastructures.stack.Stack;

import java.util.ArrayList;

public class BinaryTree<T extends Comparable<T>> {
  public Node root;

  public BinaryTree() {
  }

  public BinaryTree(Node<T> root) {
    this.root = root;
  }

  // Each depth  first traversal method should return an array of values, ordered appropriately.
  // O(N) / O(N) extra linear ds
  public Object[] iPreOrder() {
    //"i"terative solution
    if (root == null) return null;
    ArrayList<Object> list = new ArrayList();
    Stack stack = new Stack();
    stack.push(root);
    while (!stack.isEmpty()) {
      Node cur = (Node) stack.pop();
      System.out.println("cur " + cur.val);
      list.add(cur.val);
      //now check if children exist
      if (cur.right != null) stack.push(cur.right);
      if (cur.left != null) stack.push(cur.left);
    }
    return list.toArray();
  }

//  public Object[] preOrder() {
//    ArrayList values = new ArrayList();
//    preOrder(root, values);
//    return values.toArray();
//  }

  public ArrayList<T> bfs() {
    ArrayList<T> list = new ArrayList<>();
    Queue<Node<T>> q = new Queue<>();
    q.enqueue(root);
    while (!q.isEmpty()) {
      Node<T> cur = q.dequeue();
      System.out.println("bfs cur " + cur.val);
      list.add(cur.val);
      if (cur.left != null) q.enqueue(cur.left);
      if (cur.right != null) q.enqueue(cur.right);
    }

    return list;
  }

  public Object[] preOrder(Node root, ArrayList vals) {
    // node -> left -> right  PLR - DFS
    if (root == null) return null;
    System.out.println("ROOT " + root.val);
    vals.add(root.val);
    Object[] l = preOrder(root.left, vals);
    Object[] r = preOrder(root.right, vals);
    return vals.toArray();
  }

  public ArrayList inOrder() { //
    ArrayList values = new ArrayList();
    inOrder(root, values);
    return values;
  }

  public void inOrder(Node node, ArrayList values) {
    // left -> node -> right LPR
    if (node == null) return;

    inOrder(node.left, values);
    values.add(node.val);
    inOrder(node.right, values);
  }

  public Object[] postOrder() {
    ArrayList values = new ArrayList();
    postOrder(root, values);
    return values.toArray();
  }

  public void postOrder(Node node, ArrayList values) {
    // left -> right -> node LRP (root is dead last! yo )
    if (node == null) return;

    postOrder(node.left, values);
    postOrder(node.right, values);
    values.add(node.val);
    System.out.println("cur is " + node.val);
  }

  public int findMax(Node node) {
    if (node == null) return 0;
    int max = (int) node.val;
    int leftVal = findMax(node.left);
    int rightVal = findMax(node.right);
    if (leftVal > max) max = leftVal;
    if (rightVal > max) max = rightVal;
    return max;
  }

  // 17 Breadth-first Traversal. w/ Generics. LC102

  public ArrayList<T> BFS(BinaryTree<T> tree) {
    //dont't try to fight stack and qrite recursive bfs;
    Queue<Node<T>> q = new Queue();
    ArrayList<T> list = new ArrayList<>();
    if (tree.root != null) {
      q.enqueue(tree.root);
      while (q.front != null) {
        Node<T> temp = q.dequeue();
        if (temp.left != null) {
          q.enqueue(temp.left);
        }
        if (temp.right != null) {
          q.enqueue(temp.right);
        }
        list.add(temp.val);
      }
    }
    return list;
  }
}

File: tree/binaryTree.js
// Node class to represent individual nodes in the tree
const {Queue} = require("../stack-and-queue/stack-and-queue");

class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

// Binary Tree class
class BinaryTree {
    constructor() {
        this.root = null;
    }

    // Pre-order traversal
    preOrder() {
        const result = [];
        const traverse = (node) => {
            if (!node) return;
            result.push(node.value);
            traverse(node.left);
            traverse(node.right);
        };
        traverse(this.root);
        return result;
    }

    // In-order traversal
    inOrder() {
        const result = [];
        const traverse = (node) => {
            if (!node) return;
            traverse(node.left);
            result.push(node.value);
            traverse(node.right);
        };
        traverse(this.root);
        return result;
    }

    // Post-order traversal
    postOrder() {
        const result = [];
        const traverse = (node) => {
            if (!node) return;
            traverse(node.left);
            traverse(node.right);
            result.push(node.value);
        };
        traverse(this.root);
        return result;
    }

    treeMax(value) {
        let cur = this.root;
        let max = 0;
        while (cur) {
            if (cur.value > max) max = cur.value;
            if (value < cur.value) cur = cur.left;
            else cur = cur.right;
        }
        return false;
    }

}

// Binary Search Tree class extending BinaryTree
class BinarySearchTree extends BinaryTree {
    constructor() {
        super();
    }

    // Add method to add a value to the BST
    add(value) {
        const newNode = new Node(value);
        if (!this.root) {
            this.root = newNode;
            return;
        }
        let current = this.root;
        while (true) {
            if (value < current.value) {
                if (!current.left) {
                    current.left = newNode;
                    return;
                }
                current = current.left;
            } else {
                if (!current.right) {
                    current.right = newNode;
                    return;
                }
                current = current.right;
            }
        }
    }

    // Contains method to check if a value exists in the BST
    contains(value) {
        let current = this.root;
        while (current) {
            if (value === current.value) return true;
            if (value < current.value) current = current.left;
            else current = current.right;
        }
        return false;
    }



    // bfs() {
    //     let q = new Queue();
    //     let visited = new Queue()
    //     console.log(this.root.val)
    //     while (this.root != null) {
    //         q.enqueue(this.root.left.val);
    //         q.enqueue(this.root.right.val);
    //         visited.enqueue(q.dequeue())
    //         visited.enqueue(q.dequeue())
    //         this.root = root.left.next;
    //     }
    // }
}

const bst = new BinarySearchTree();

// Add nodes to the tree
bst.add(10);
bst.add(5);
bst.add(15);
bst.add(3);
bst.add(7);
bst.add(12);
bst.add(18);

console.log(bst.bfs())

module.exports = BinarySearchTree;
